shells/pipes-shell/web/cache-mgr.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+const ARCS_CACHE_MANAGER = './cache-mgr-sw.js';
+
+// Arcs Cache Manager
+(async function _arcsCacheManager() {
+  if ('serviceWorker' in navigator) {
+    // Service Worker supports ONLY https:// and http://localhost.
+    // Only https:// is accepted at this moment since local socket test
+    // servers are not supported on devices.
+    if (location.protocol !== 'https:') return;
+
+    // Only registering the new cache manager service worker instance when
+    // requested in 'https' protocol and with ?use-cache url parameter,
+    // otherwise unregistering the existing service workers at this scope.
+    const params = new URLSearchParams(window.location.search);
+    if (!params.has('use-cache')) {
+      navigator.serviceWorker.getRegistrations().then(function(regs) {
+        for (const reg of regs) {
+          console.log('unregistering Arcs Cache Manager: ' + reg.scope);
+          reg.unregister();
+        }
+      }).catch(function(err) {
+        console.log('Arcs Cache Manager unregistration failed: ' + err);
+      });
+      return;
+    }
+
+    navigator.serviceWorker.register(ARCS_CACHE_MANAGER)
+      .then(reg => {
+        console.log(`Arcs Cache Manager registered, scope: ${reg.scope}`);
+      })
+      .catch(err => {
```

Comment:
Instead of promise chaining like this, it's generally reads a little better to use `await` inside an async function, e.g.:

```
async function startCacheManager() {
  ...
  await navigator.serviceWorker.register(ARCS_CACHE_MANAGER);
  console.log(`Arcs Cache Manager registered, scope: ${reg.scope}`);
```

Then at the bottom of the file, you can invoke it like so:

```
startCacheManager().catch(e => console.error(...));
```

Reply:
Thanks, I thought the chain is not nested (just single level) thus I didn't break it into separate await + async function.

Reply:
I'll still keep the single level of chain in then/catch for now to make the code short and gathered. Then change it to async/await or Promise.all when multiple conditions need to be met.